docs/cache-server.md

```diff
@@ -0,0 +1,125 @@
+### Purpose
+
+The primary purpose of the cache server is to support cross-shard communication.
+Shards need to communicate with each other in order to enable, among other things, migration and replication features.
+Direct shard-to-shard communication is not feasible in a larger setup.
+The web of connections would be hard to reason about, maintain and troubleshoot. 
+Thus, the cache server serves as a central place for shards to store and read common data.
+
+### High-level overview
+
+The cache server is a regular, Kubernetes-style CRUD API server with support of LIST/WATCH semantics. 
+Conceptually it supports two modes of operations. 
+The first mode is in which a shard gets its private space for storing its own data. 
+The second mode is in which a shard can read other shards' data.
+
+The first mode of operation is implemented as a controller that runs within a shard. 
+It maintains an informer for `LIST/WATCH`ing its own resources. 
+It compares received data to its local copy and makes sure both copies are consistent.
+
+The second mode of operation is implemented as a separate informer within a shard. 
```

Comment:
Also here - this is a controller, on some shard, holding other shards' data in an in-memory cache using informer, pulling data from the central cache